k8s/pkg/apis/longhorn/v1beta2/objectstore.go

```diff
@@ -0,0 +1,175 @@
+package v1beta2
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ObjectStoreState string
+
+const (
+	ObjectStoreStateUnknown  = ObjectStoreState("Unknown")
+	ObjectStoreStateStarting = ObjectStoreState("Starting")
+	ObjectStoreStateRunning  = ObjectStoreState("Running")
+	ObjectStoreStateStopping = ObjectStoreState("Stopping")
+	ObjectStoreStateError    = ObjectStoreState("Error")
+)
+
+type ObjectStoreStatus struct {
+	// The state of the object store as observed by the object store controller.
+	//
+	// The object store implemets a state machine with this, beginning at
+	// "unknown". Once the object store controller detects the new object store
+	// and begins resource creation the state is transformed to "starting". The
+	// object store controller observes the resources and once all resources are
+	// up and ready the state transitions to "running".
+	// The state remains in "running" until the object store is deleted, at
+	// which point the controller will clean up the associated resources.
+	//
+	//  │ object store created
+	//  │
+	//  └──►┌───────────┐
+	//      │ unknown   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller creates resources
+	//  │
+	//  └──►┌───────────┐
+	//      │ starting  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller detects all resources ready
+	//  │
+	//  └──►┌───────────┐
+	//      │ running   │
+	//  ┌── └───────────┘
+	//  │
+	//  │ object endpoint is marked for deletion
+	//  │
+	//  └──►┌───────────┐
+	//      │ stopping  │
+	//  ┌── └───────────┘
+	//  │
+	//  │ controller waits for dependent resources to disappear before removing
+	//  │ the finalizer, thereby letting the object store be deleted
```

Comment:
> This ACSII comment graph is cool! Did you use any tool to create it?
Thanks. I used asciiflow.com. The diagram is unfortunately still missing the `error` state.

> it is necessary to allow the user to be able to sometimes voluntarily stop an ObjectStore
If this is a valid use case we should support it. Adding an additional "stopped" state makes sense for this purpose.

The purpose of the `stopping` state here (or `deleting` state as you proposed) is to keep the ObjectStore K8s API object around until all resources have been cleaned up. This allows the user to monitor the cleanup and detect if anything goes wrong, so they can intervene, otherwise we might silently leak K8s resource objects.
I'd prefer use K8s terminology analog to Pods and call that state `terminating` but that's a minor detail.

Reply:
Sounds good to me:
* Repurpose the `stopping` and `stopped` state for a valid use-case that user voluntary stop/start an ObjectStore by setting spec.DesireState
* When the deletionTimeStamp is set, we can enter the `terminating` state and wait for all dependency resource to be cleaned up before remove the `longhorn.io` finalizer and allow the ObjectStore CR to be gone